module1-introduction-to-backend/r3.4-api-best-practices/README.md

```diff
@@ -0,0 +1,232 @@
+  const userExists = users.find(user => user.email === email);
+  if (userExists) {
+    return res.status(400).json({ error: 'User already exists' })
+  }
+  res.json(req.body);
+});
+
+
+app.listen(3000, () => console.log('server started'));
+```
+
+6. **Allow filtering, sorting, and pagination**
+The databases behind a REST API can get very large. Sometimes, there’s so much data that it shouldn’t be returned all at once because it’s way too slow or will bring down our systems. Therefore, we need ways to filter items. We also need ways to paginate data so that we only return a few results at a time. We don’t want to tie up resources for too long by trying to get all the requested data at once. Filtering and pagination both increase performance by reducing the usage of server resources.
+
+Here’s a small example where an API can accept a query string with various query parameters to let us filter out items by their fields:
+```
+const express = require('express');
+const bodyParser = require('body-parser');
+
+const app = express();
+
+// employees data in a database
+const employees = [
+  { firstName: 'Jane', lastName: 'Smith', age: 20 },
+  //...
+  { firstName: 'John', lastName: 'Smith', age: 30 },
+  { firstName: 'Mary', lastName: 'Green', age: 50 },
+]
+
+app.use(bodyParser.json());
+
+app.get('/employees', (req, res) => {
+  const { firstName, lastName, age } = req.query;
+  let results = [...employees];
+  if (firstName) {
+    results = results.filter(r => r.firstName === firstName);
+  }
+
+  if (lastName) {
+    results = results.filter(r => r.lastName === lastName);
+  }
+
+  if (age) {
+    results = results.filter(r => +r.age === +age);
+  }
+  res.json(results);
+});
+
+app.listen(3000, () => console.log('server started'));
+```
+
+7. **Maintain good security practices**
+Most communication between client and server should be private since we often send and receive private information. Therefore, using SSL/TLS for security is a must. A SSL certificate isn’t too difficult to load onto a server and the cost is free or very low. There’s no reason not to make our REST APIs communicate over secure channels instead of in the open.
+
+People shouldn’t be able to access more information that they requested. For example, a normal user shouldn’t be able to access information of another user. They also shouldn’t be able to access data of admins. To enforce the principle of least privilege, we need to add role checks either for a single role, or have more granular roles for each user.
+
+If we choose to group users into a few roles, then the roles should have the permissions that cover all they need and no more. If we have more granular permissions for each feature that users have access to, then we have to make sure that admins can add and remove those features from each user accordingly. Also, we need to add some preset roles that can be applied to a group users so that we don’t have to do that for every user manually.
```

Comment:
consider concrete example or further reading, for example something from ehre
https://www.google.com/search?q=express+role+based+authentication+framework

Reply:
These will be covered in the authentication module, so not going into examples here.

Also this reading includes this disclaimer in the beginning:
Although some of these suggestions may require going through more modules of the course before implementation, consider this as a guide that you can come back and refer to at any time as you progress through the course.